web/src/lib.rs

```diff
@@ -1,60 +1,291 @@
+        open_db_request.set_onerror(None);
+        open_db_request.set_onsuccess(None);
+
+        drop(on_upgrade_needed_closure);
+        drop(on_error_closure);
+        drop(on_success_closure);
+
+        let db_js = open_db_request.result().unwrap();
+        let db = IdbDatabase::from(db_js);
+
+        Ok(db)
+    }
+
+    async fn transaction<B>(&self, mode: IdbTransactionMode, block: B) -> Result<JsValue, ()> where B: Fn(IdbObjectStore) -> IdbRequest {
+        let db = IndexedDBStore::open_db(&self.name).await.unwrap();
+        let transaction = db.transaction_with_str_and_mode("entries", mode).unwrap();
+        let object_store = transaction.object_store("entries").unwrap();
+        let request = block(object_store);
+
+        let (on_error_sender, on_error_receiver) = oneshot::channel::<Result<(), ()>>();
+        let (on_success_sender, on_success_receiver) = oneshot::channel::<Result<(), ()>>();
+
+        let on_error_closure: Closure<dyn FnMut(_)> = Closure::once(move |_event: Event| {
+            info!("Error occured while trying to get value for key.");
+            // TODO: extract proper error object.
+            on_error_sender.send(Err(())).unwrap();
+        });
+
+        let on_success_closure: Closure<dyn FnMut(_)> = Closure::once(move |_event: Event| {
+            on_success_sender.send(Ok(())).unwrap();
+        });
+
+        request.set_onerror(on_error_closure.as_ref().dyn_ref());
+        request.set_onsuccess(on_success_closure.as_ref().dyn_ref());
+
+        let on_error_receiver_fused = on_error_receiver.fuse();
+        let on_success_receiver_fused = on_success_receiver.fuse();
+
+        pin_mut!(on_error_receiver_fused, on_success_receiver_fused);
+
+        let result: Result<(), ()> = select! {
+            // TODO: pass proper error.
+            _ = on_error_receiver_fused => Err(()),
+            _ = on_success_receiver_fused => Ok(()),
+        };
+
+        result.expect("An error occured while trying to get value for key.");
+        db.close();
+
+        request.set_onerror(None);
+        request.set_onsuccess(None);
+
+        drop(on_error_closure);
+        drop(on_success_closure);
+
+        let js_value = request.result().unwrap();
+        Ok(js_value)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_example() {
+        // Nothing here yet, all good.
+    }
+}
+
+#[wasm_bindgen]
+// NOTE: can't use term `Type` here, conflicts with keyword `type` and
+// leads to weird variable names. Therefore "variant" it is.
+pub enum JSStoreVariant {
+    InMemory,
+    IndexedDB
+}
+
 #[wasm_bindgen]
 pub struct JSStore {
-    store: Rc<RefCell<InMemoryStore>>,
+    // TODO: check for thread safety, probably need a replacement for
+    // RefCell, which isn't thread-safe. Also, we probably won't need
+    // dynamic dispatch since we pick the appropriate implementation
+    // via rollup.js (statically).
+    store: Arc<RefCell<dyn Store>>,
```

Comment:
I wonder if there really is a need to be concerned about thread safety. While wasm can be multi-threaded, I believe that since js is single threaded and each web worker is in an isolated environment where the wasm would be compiled again -- there is not thread safety to worry about.

Reply:
Yep, that's exactly what I wasn't sure about, including the web workers. So we can even replace ARC with RC then, right?